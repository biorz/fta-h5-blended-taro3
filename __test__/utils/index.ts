import detectPort from 'detect-port'
import { createFsFromVolume, IFs, Volume } from 'memfs'
import joinPath from 'memory-fs/lib/join'
import * as path from 'path'
import * as webpack from 'webpack'
import WebpackDevServer from 'webpack-dev-server'

console.log('WebpackDevServer', WebpackDevServer)

export async function compile(config: webpack.Configuration) {
  process.env.TARO_ENV = 'h5'

  const appPath = path.resolve(__dirname, '../mock')

  process.chdir(appPath)

  const stats = await run(config)
  return { stats, config }
}

function run(webpackConfig: webpack.Configuration): Promise<webpack.Stats | undefined> {
  const compiler = webpack.default(webpackConfig)
  const fs = createFsFromVolume(new Volume())
  const ensuredFs = ensureWebpackMemoryFs(fs)

  compiler.outputFileSystem = ensuredFs

  return new Promise((resolve, reject) => {
    compiler.run((err, stats) => {
      if (err ?? stats?.hasErrors()) {
        const error = err ?? stats!.toJson().errors
        reject(error)
      } else {
        resolve(stats)
      }
    })
  })
}

interface EnsuredFs extends IFs {
  join: () => string
}

function ensureWebpackMemoryFs(fs: IFs): EnsuredFs {
  const newFs: EnsuredFs = Object.create(fs)
  newFs.join = joinPath

  return newFs
}

export function getOutput(stats, config: webpack.Configuration) {
  const fs: IFs = stats.compilation.compiler.outputFileSystem

  const files = readDir(fs, config.output.path)
  const output = files.reduce((content, file) => {
    return `${content}
/** filePath: ${file} **/
${fs.readFileSync(file)}
`
  }, '')
  return output
}

function readDir(fs: IFs, dir: string) {
  let files: string[] = []
  const list = fs.readdirSync(dir)
  list.forEach((item) => {
    const filePath = path.join(dir, item)
    const stat = fs.statSync(filePath)
    if (stat.isDirectory()) {
      files = files.concat(readDir(fs, filePath))
    } else {
      files.push(filePath)
    }
  })
  return files
}

export const devServer = async (
  webpackConfig: webpack.Configuration
): Promise<webpack.Stats | undefined> => {
  const port = await detectPort(8080)
  const devServerOptions = {
    host: 'localhost',
    port,
  }

  return new Promise((resolve, reject) => {
    WebpackDevServer.addDevServerEntrypoints(webpackConfig, devServerOptions)
    const compiler = webpack.default(webpackConfig) as webpack.Compiler
    const server = new WebpackDevServer(compiler, devServerOptions)
    compiler.hooks.done.tap('taroBuildDone', (stats) => {
      resolve(stats)
    })

    compiler.hooks.failed.tap('taroBuildDone', (error) => {
      reject(error)
    })

    server.listen(devServerOptions.port, devServerOptions.host as string, (err) => {
      if (err) {
        console.log(err)
      }
    })
  })
}
