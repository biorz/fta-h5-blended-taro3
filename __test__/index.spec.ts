import webpackChain from 'webpack-chain'
import h5BlendedTaro3 from '../src'
import webpackConfig from './mock/webpack.config'
import { compile, devServer, getOutput } from './utils'

describe('chain', () => {
  jest.setTimeout(1000000)

  test('modify webpack config', () => {
    const newConfig = h5BlendedTaro3(webpackConfig)

    expect(newConfig).toMatchSnapshot()
  })

  test('modify webpack chain', () => {
    const chain = h5BlendedTaro3(new webpackChain())

    expect(chain.toConfig()).toMatchSnapshot()
  })

  test('inject entry on build', async () => {
    const chain = h5BlendedTaro3(webpackConfig)

    const { stats, config } = await compile(chain)
    const assets = stats?.toJson().assets || []
    expect(assets.length).toMatchSnapshot()
    const output = getOutput(stats, config)
    expect(output).toEqual(expect.stringContaining('defineCustomElements'))
  })

  test('inject entry on dev server', async () => {
    const config = h5BlendedTaro3(webpackConfig)

    const stats = await devServer(config)
    const assets = stats?.toJson().assets || []
    expect(assets.length).toMatchSnapshot()
    const output = getOutput(stats, config)
    expect(output).toEqual(expect.stringContaining('defineCustomElements'))
  })

  test('entries of object type', async () => {
    const config = h5BlendedTaro3({
      ...webpackConfig,
      entry: {
        app: './src/app.tsx',
      },
    })

    const stats = await devServer(config)
    const assets = stats?.toJson().assets || []
    expect(assets.length).toMatchSnapshot()
    const output = getOutput(stats, config)
    expect(output).toEqual(expect.stringContaining('defineCustomElements'))
  })
})
