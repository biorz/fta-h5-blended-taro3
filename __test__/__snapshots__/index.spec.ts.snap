// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`chain entries of object type 1`] = `39`;

exports[`chain inject entry on build 1`] = `39`;

exports[`chain inject entry on dev server 1`] = `39`;

exports[`chain modify webpack chain 1`] = `
Object {
  "module": Object {
    "rules": Array [
      Object {
        "include": Array [
          /@tarojs/,
        ],
        "test": /\\\\\\.js\\|\\\\\\.ts\\$/,
        "use": Array [
          Object {
            "loader": "babel-loader",
          },
        ],
      },
    ],
  },
  "plugins": Array [
    DefinePlugin {
      "definitions": Object {
        "ENABLE_ADJACENT_HTML": "false",
        "ENABLE_CLONE_NODE": "false",
        "ENABLE_CONTAINS": "false",
        "ENABLE_INNER_HTML": "false",
        "ENABLE_MUTATION_OBSERVER": "false",
        "ENABLE_SIZE_APIS": "false",
        "ENABLE_TEMPLATE_CONTENT": "false",
        "process.env.TARO_ENV": "\\"h5\\"",
      },
    },
    MainPlugin {
      "options": Object {
        "designRadio": Object {
          "750": 1,
        },
        "designWidth": 750,
        "deviceRatio": Object {},
      },
    },
  ],
  "resolve": Object {
    "alias": Object {
      "@tarojs/components$": "@tarojs/components/dist-h5/react",
      "@tarojs/taro": "@tarojs/taro-h5",
    },
    "mainFields": Array [
      "main:h5",
      "browser",
      "module",
      "jsnext:main",
      "main",
    ],
  },
}
`;

exports[`chain modify webpack config 1`] = `
Object {
  "context": "/Users/biorz/workspace/fta-view/packages/config/h5-blended-taro3/__test__/mock",
  "entry": Array [
    "/Users/biorz/workspace/fta-view/packages/config/h5-blended-taro3/__test__/mock/src/app.tsx",
  ],
  "externals": Array [
    "react",
  ],
  "mode": "development",
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /node_modules/,
        "include": Array [
          "/Users/biorz/workspace/fta-view/packages/config/h5-blended-taro3/__test__/mock/src",
        ],
        "loader": "babel-loader",
        "test": /\\\\\\.\\(js\\|mjs\\|jsx\\|ts\\|tsx\\)\\$/,
      },
      Object {
        "test": /\\\\\\.css\\$/,
        "use": Array [
          "style-loader",
          "css-loader",
          "postcss-loader",
        ],
      },
      Object {
        "include": Array [
          /@tarojs/,
        ],
        "test": /\\\\\\.js\\|\\\\\\.ts\\$/,
        "use": Array [
          Object {
            "loader": "babel-loader",
          },
        ],
      },
    ],
  },
  "output": Object {
    "path": "/Users/biorz/workspace/fta-view/packages/config/h5-blended-taro3/__test__/mock/dist",
  },
  "plugins": Array [
    DefinePlugin {
      "definitions": Object {
        "process.env": Object {
          "CODE_ENV": undefined,
          "NODE_ENV": "\\"test\\"",
          "USEMOCK": undefined,
        },
      },
    },
    DefinePlugin {
      "definitions": Object {
        "ENABLE_ADJACENT_HTML": "false",
        "ENABLE_CLONE_NODE": "false",
        "ENABLE_CONTAINS": "false",
        "ENABLE_INNER_HTML": "false",
        "ENABLE_MUTATION_OBSERVER": "false",
        "ENABLE_SIZE_APIS": "false",
        "ENABLE_TEMPLATE_CONTENT": "false",
        "process.env.TARO_ENV": "\\"h5\\"",
      },
    },
    MainPlugin {
      "options": Object {
        "designRadio": Object {
          "750": 1,
        },
        "designWidth": 750,
        "deviceRatio": Object {},
      },
    },
  ],
  "resolve": Object {
    "alias": Object {
      "@mw": "/Users/biorz/workspace/fta-view/packages/config/h5-blended-taro3/__test__/src/assets",
      "@tarojs/components$": "@tarojs/components/dist-h5/react",
      "@tarojs/taro": "@tarojs/taro-h5",
      "rider": "/Users/biorz/workspace/fta-view/packages/config/h5-blended-taro3/__test__/node_modules/rider/lib/rider/index.styl",
      "src": "/Users/biorz/workspace/fta-view/packages/config/h5-blended-taro3/__test__/src",
    },
    "extensions": Array [
      "*",
      ".js",
      ".jsx",
      ".json",
      ".ts",
      ".tsx",
    ],
    "mainFields": Array [
      "main:h5",
      "browser",
      "module",
      "jsnext:main",
      "main",
      "main:brower",
    ],
  },
}
`;
